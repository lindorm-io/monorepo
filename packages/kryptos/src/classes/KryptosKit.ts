import { B64 } from "@lindorm/b64";
import { KryptosError } from "../errors";
import {
  IKryptos,
  IKryptosEc,
  IKryptosOct,
  IKryptosOkp,
  IKryptosRsa,
} from "../interfaces";
import {
  KryptosAuto,
  KryptosDB,
  KryptosFormat,
  KryptosFrom,
  KryptosFromBuffer,
  KryptosFromDb,
  KryptosFromJwk,
  KryptosFromString,
  KryptosGenerateEcEnc,
  KryptosGenerateEcSig,
  KryptosGenerateOctEnc,
  KryptosGenerateOctSig,
  KryptosGenerateOkpEnc,
  KryptosGenerateOkpSig,
  KryptosGenerateRsaEnc,
  KryptosGenerateRsaSig,
  KryptosLike,
} from "../types";
import { KryptosGenerate } from "../types/private";
import {
  autoGenerateConfig,
  calculateKeyOps,
  fromOptions,
  generateKey,
  isB64,
  isDer,
  isJwk,
  isPem,
} from "../utils/private";
import { Kryptos } from "./Kryptos";

type Env = {
  import(string: string): IKryptos;
  export(kryptos: IKryptos): string;
};

type From = {
  auto(options: KryptosFrom): IKryptos;
  b64(options: KryptosFromString): IKryptos;
  db(options: KryptosDB): IKryptos;
  der(options: KryptosFromBuffer): IKryptos;
  jwk(options: KryptosFromJwk): IKryptos;
  pem(options: KryptosFromString): IKryptos;
  utf(options: KryptosFromString): IKryptos;
};

type Encryption = {
  ec(options: KryptosGenerateEcEnc): IKryptos;
  oct(options: KryptosGenerateOctEnc): IKryptos;
  okp(options: KryptosGenerateOkpEnc): IKryptos;
  rsa(options: KryptosGenerateRsaEnc): IKryptos;
};

type Signature = {
  ec(options: KryptosGenerateEcSig): IKryptos;
  oct(options: KryptosGenerateOctSig): IKryptos;
  okp(options: KryptosGenerateOkpSig): IKryptos;
  rsa(options: KryptosGenerateRsaSig): IKryptos;
};

type Generate = {
  auto(options: KryptosAuto): IKryptos;
  enc: Encryption;
  sig: Signature;
};

export class KryptosKit {
  // clone

  public static clone(kryptos: IKryptos, overwrite: KryptosLike = {}): IKryptos {
    return new Kryptos({ ...kryptos.toJSON(), ...overwrite, ...kryptos.export("der") });
  }

  // env

  public static get env(): Env {
    return {
      import: KryptosKit.envImport,
      export: KryptosKit.envExport,
    };
  }

  // from

  public static get from(): From {
    return {
      auto: KryptosKit.fromAuto,
      b64: KryptosKit.fromB64,
      db: KryptosKit.fromDb,
      der: KryptosKit.fromDer,
      jwk: KryptosKit.fromJwk,
      pem: KryptosKit.fromPem,
      utf: KryptosKit.fromUtf,
    };
  }

  // is

  public static isEc(kryptos: KryptosLike): kryptos is IKryptosEc {
    return kryptos instanceof Kryptos && kryptos.type === "EC" && Boolean(kryptos.curve);
  }

  public static isOct(kryptos: KryptosLike): kryptos is IKryptosOct {
    return kryptos instanceof Kryptos && kryptos.type === "oct" && !kryptos.curve;
  }

  public static isOkp(kryptos: KryptosLike): kryptos is IKryptosOkp {
    return kryptos instanceof Kryptos && kryptos.type === "OKP" && Boolean(kryptos.curve);
  }

  public static isRsa(kryptos: KryptosLike): kryptos is IKryptosRsa {
    return kryptos instanceof Kryptos && kryptos.type === "RSA" && !kryptos.curve;
  }

  // generate

  public static get generate(): Generate {
    return {
      auto: this.generateAuto,
      enc: this.encryption,
      sig: this.signature,
    };
  }

  // private env

  private static envImport(string: string): IKryptos {
    if (!string.startsWith("kryptos:")) {
      throw new KryptosError("Invalid kryptos string");
    }

    const [_, jwk] = string.split(":");

    return KryptosKit.fromJwk(JSON.parse(B64.decode(jwk, "b64u")));
  }

  private static envExport(kryptos: IKryptos): string {
    return kryptos.toString();
  }

  // private from

  private static fromAuto(options: KryptosFrom): IKryptos {
    if (isB64(options)) return KryptosKit.fromB64(options);
    if (isDer(options)) return KryptosKit.fromDer(options);
    if (isJwk(options)) return KryptosKit.fromJwk(options);
    if (isPem(options)) return KryptosKit.fromPem(options);

    throw new KryptosError("Unexpected key format");
  }

  private static fromB64(options: KryptosFromString): IKryptos {
    return KryptosKit.fromKryptos("b64", options);
  }

  private static fromDb(options: KryptosFromDb): IKryptos {
    const { curve, encryption, privateKey, publicKey, ...rest } = options;

    return KryptosKit.fromKryptos("b64", {
      ...rest,
      curve: curve ?? undefined,
      encryption: encryption ?? undefined,
      privateKey: privateKey ?? undefined,
      publicKey: publicKey ?? undefined,
    });
  }

  private static fromDer(options: KryptosFromBuffer): IKryptos {
    return KryptosKit.fromKryptos("der", options);
  }

  private static fromJwk(options: KryptosFromJwk): IKryptos {
    return KryptosKit.fromKryptos("jwk", options);
  }

  private static fromPem(options: KryptosFromString): IKryptos {
    return KryptosKit.fromKryptos("pem", options);
  }

  private static fromUtf(options: KryptosFromString): IKryptos {
    return KryptosKit.fromKryptos("utf", options);
  }

  private static fromKryptos(format: KryptosFormat, arg: KryptosFrom): Kryptos {
    const options = fromOptions(format, arg);

    if (!options.algorithm) {
      throw new KryptosError("Algorithm is required");
    }
    if (!options.type) {
      throw new KryptosError("Type is required");
    }
    if (!options.use) {
      throw new KryptosError("Use is required");
    }

    return new Kryptos({
      ...options,
      operations: options.operations?.length
        ? options.operations
        : calculateKeyOps(options),
    });
  }

  // private generate

  private static get encryption(): Encryption {
    return {
      ec: this.generateEcEnc,
      oct: this.generateOctEnc,
      okp: this.generateOkpEnc,
      rsa: this.generateRsaEnc,
    };
  }

  private static get signature(): Signature {
    return {
      ec: this.GenerateEcSig,
      oct: this.generateOctSig,
      okp: this.generateOkpSig,
      rsa: this.generateRsaSig,
    };
  }

  private static generateAuto(options: KryptosAuto): IKryptos {
    const config = autoGenerateConfig(options.algorithm);
    const operations = calculateKeyOps(config);

    if (config.purpose && ["cookie", "session"].includes(config.purpose)) {
      config.hidden = config.hidden ?? true;
    }

    const generate: KryptosGenerate = {
      ...config,
      ...options,
      operations,
    };

    return new Kryptos({ ...generate, ...generateKey(generate) });
  }

  private static generateEcEnc(options: KryptosGenerateEcEnc): IKryptos {
    const generate: KryptosGenerate = {
      ...options,
      type: "EC",
      use: "enc",
    };
    return KryptosKit.generateKryptos(generate);
  }

  private static GenerateEcSig(options: KryptosGenerateEcSig): IKryptos {
    const generate: KryptosGenerate = {
      ...options,
      type: "EC",
      use: "sig",
    };
    return KryptosKit.generateKryptos(generate);
  }

  private static generateOctEnc(options: KryptosGenerateOctEnc): IKryptos {
    const generate: KryptosGenerate = {
      ...options,
      type: "oct",
      use: "enc",
    };
    return KryptosKit.generateKryptos(generate);
  }

  private static generateOctSig(options: KryptosGenerateOctSig): IKryptos {
    const generate: KryptosGenerate = {
      ...options,
      type: "oct",
      use: "sig",
    };
    return KryptosKit.generateKryptos(generate);
  }

  private static generateOkpEnc(options: KryptosGenerateOkpEnc): IKryptos {
    const generate: KryptosGenerate = {
      ...options,
      type: "OKP",
      use: "enc",
    };
    return KryptosKit.generateKryptos(generate);
  }

  private static generateOkpSig(options: KryptosGenerateOkpSig): IKryptos {
    const generate: KryptosGenerate = {
      ...options,
      type: "OKP",
      use: "sig",
    };
    return KryptosKit.generateKryptos(generate);
  }

  private static generateRsaEnc(options: KryptosGenerateRsaEnc): IKryptos {
    const generate: KryptosGenerate = {
      ...options,
      type: "RSA",
      use: "enc",
    };
    return KryptosKit.generateKryptos(generate);
  }

  private static generateRsaSig(options: KryptosGenerateRsaSig): IKryptos {
    const generate: KryptosGenerate = {
      ...options,
      type: "RSA",
      use: "sig",
    };
    return KryptosKit.generateKryptos(generate);
  }

  private static generateKryptos(generate: KryptosGenerate): IKryptos {
    return new Kryptos({
      ...generate,
      encryption: generate.use === "enc" ? (generate.encryption ?? "A256GCM") : null,
      operations: generate.operations?.length
        ? generate.operations
        : calculateKeyOps(generate),
      ...generateKey(generate),
    });
  }
}
