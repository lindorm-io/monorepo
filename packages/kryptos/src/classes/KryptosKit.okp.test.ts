import MockDate from "mockdate";
import {
  TEST_OKP_KEY_B64,
  TEST_OKP_KEY_JWK,
  TEST_OKP_KEY_PEM,
} from "../__fixtures__/okp-keys";
import { KryptosKit } from "./KryptosKit";

const MockedDate = new Date("2024-01-01T08:00:00.000Z");
MockDate.set(MockedDate.toISOString());

jest.mock("crypto", () => ({
  ...jest.requireActual("crypto"),
  randomUUID: jest.fn().mockReturnValue("6e6f84b0-e125-5e3f-90ae-c65269668d98"),
}));

describe("KryptosKit (OKP)", () => {
  describe("clone", () => {
    test("should clone", () => {
      const kryptos = KryptosKit.from.auto(TEST_OKP_KEY_B64);
      const cloned = KryptosKit.clone(kryptos);

      expect(cloned.toJSON()).toEqual(kryptos.toJSON());
      expect(cloned.toJWK()).toEqual(kryptos.toJWK());
      expect(cloned.export("b64")).toEqual(kryptos.export("b64"));
    });
  });

  describe("from", () => {
    describe("auto", () => {
      test("b64", () => {
        const kryptos = KryptosKit.from.auto(TEST_OKP_KEY_B64);

        expect(kryptos.toJSON()).toMatchSnapshot();
        expect(kryptos.toJWK()).toMatchSnapshot();
        expect(kryptos.export("b64")).toMatchSnapshot();
      });

      test("jwk", () => {
        const kryptos = KryptosKit.from.auto(TEST_OKP_KEY_JWK);

        expect(kryptos.toJSON()).toMatchSnapshot();
        expect(kryptos.toJWK()).toMatchSnapshot();
        expect(kryptos.export("jwk")).toMatchSnapshot();
      });

      test("pem", () => {
        const kryptos = KryptosKit.from.auto(TEST_OKP_KEY_PEM);

        expect(kryptos.toJSON()).toMatchSnapshot();
        expect(kryptos.toJWK()).toMatchSnapshot();
        expect(kryptos.export("pem")).toMatchSnapshot();
      });
    });

    test("b64", () => {
      const kryptos = KryptosKit.from.b64(TEST_OKP_KEY_B64);

      expect(kryptos.toJSON()).toMatchSnapshot();
      expect(kryptos.toJWK()).toMatchSnapshot();
      expect(kryptos.export("b64")).toMatchSnapshot();
    });

    test("jwk", () => {
      const kryptos = KryptosKit.from.jwk(TEST_OKP_KEY_JWK);

      expect(kryptos.toJSON()).toMatchSnapshot();
      expect(kryptos.toJWK()).toMatchSnapshot();
      expect(kryptos.export("jwk")).toMatchSnapshot();
    });

    test("pem", () => {
      const kryptos = KryptosKit.from.pem(TEST_OKP_KEY_PEM);

      expect(kryptos.toJSON()).toMatchSnapshot();
      expect(kryptos.toJWK()).toMatchSnapshot();
      expect(kryptos.export("pem")).toMatchSnapshot();
    });
  });

  describe("is", () => {
    test("isEc", () => {
      const kryptos = KryptosKit.from.b64(TEST_OKP_KEY_B64);

      expect(KryptosKit.isEc(kryptos)).toBe(false);
    });

    test("isOct", () => {
      const kryptos = KryptosKit.from.b64(TEST_OKP_KEY_B64);

      expect(KryptosKit.isOct(kryptos)).toBe(false);
    });

    test("isOkp", () => {
      const kryptos = KryptosKit.from.b64(TEST_OKP_KEY_B64);

      expect(KryptosKit.isOkp(kryptos)).toBe(true);
    });

    test("isRsa", () => {
      const kryptos = KryptosKit.from.b64(TEST_OKP_KEY_B64);

      expect(KryptosKit.isRsa(kryptos)).toBe(false);
    });
  });

  describe("make", () => {
    test("auto", () => {
      const kryptos = KryptosKit.make.auto({
        algorithm: "ECDH-ES+A128GCMKW",
      });

      expect(kryptos.toJSON()).toMatchSnapshot();
      expect(kryptos.export("b64")).toEqual({
        algorithm: "ECDH-ES+A128GCMKW",
        curve: "P-256",
        privateKey: expect.any(String),
        publicKey: expect.any(String),
        type: "EC",
        use: "enc",
      });
    });

    test("enc", () => {
      const kryptos = KryptosKit.make.enc.okp({
        algorithm: "ECDH-ES",
        encryption: "A256GCM",
      });

      expect(kryptos.toJSON()).toMatchSnapshot();
      expect(kryptos.export("b64")).toEqual({
        algorithm: "ECDH-ES",
        curve: "X25519",
        privateKey: expect.any(String),
        publicKey: expect.any(String),
        type: "OKP",
        use: "enc",
      });
    });

    test("sig", () => {
      const kryptos = KryptosKit.make.sig.okp({
        algorithm: "EdDSA",
      });

      expect(kryptos.toJSON()).toMatchSnapshot();
      expect(kryptos.export("b64")).toEqual({
        algorithm: "EdDSA",
        curve: "Ed448",
        privateKey: expect.any(String),
        publicKey: expect.any(String),
        type: "OKP",
        use: "sig",
      });
    });
  });
});
