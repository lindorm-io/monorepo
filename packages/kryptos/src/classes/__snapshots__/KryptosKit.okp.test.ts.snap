// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KryptosKit (OKP) from auto b64 1`] = `
{
  "algorithm": "EdDSA",
  "createdAt": 2024-01-01T08:00:00.000Z,
  "curve": "Ed25519",
  "encryption": null,
  "expiresAt": null,
  "expiresIn": -1,
  "hasPrivateKey": true,
  "hasPublicKey": true,
  "hidden": false,
  "id": "ff2a448e-8eb0-59a8-8fc6-0ce23ad07e8f",
  "isActive": true,
  "isExpired": false,
  "isExternal": false,
  "issuer": null,
  "jwksUri": null,
  "modulus": null,
  "notBefore": 2024-01-01T08:00:00.000Z,
  "operations": [
    "sign",
    "verify",
  ],
  "ownerId": null,
  "purpose": null,
  "type": "OKP",
  "updatedAt": 2024-01-01T08:00:00.000Z,
  "use": "sig",
}
`;

exports[`KryptosKit (OKP) from auto b64 2`] = `
{
  "alg": "EdDSA",
  "crv": "Ed25519",
  "iat": 1704096000,
  "key_ops": [
    "sign",
    "verify",
  ],
  "kid": "ff2a448e-8eb0-59a8-8fc6-0ce23ad07e8f",
  "kty": "OKP",
  "nbf": 1704096000,
  "uat": 1704096000,
  "use": "sig",
  "x": "GRCwCA6lChosFGMQwxGiHCdzblfvCz0FNiRtTnm1qqc",
}
`;

exports[`KryptosKit (OKP) from auto b64 3`] = `
{
  "algorithm": "EdDSA",
  "curve": "Ed25519",
  "id": "ff2a448e-8eb0-59a8-8fc6-0ce23ad07e8f",
  "privateKey": "MC4CAQAwBQYDK2VwBCIEIBwKJlvoh1ngd9LRd7dtvGOSqW4uZamdvIu0ABD2AkxL",
  "publicKey": "MCowBQYDK2VwAyEAGRCwCA6lChosFGMQwxGiHCdzblfvCz0FNiRtTnm1qqc",
  "type": "OKP",
  "use": "sig",
}
`;

exports[`KryptosKit (OKP) from auto jwk 1`] = `
{
  "algorithm": "EdDSA",
  "createdAt": 2024-01-01T08:00:00.000Z,
  "curve": "Ed25519",
  "encryption": null,
  "expiresAt": null,
  "expiresIn": -1,
  "hasPrivateKey": true,
  "hasPublicKey": true,
  "hidden": false,
  "id": "6fc93fb6-d40e-52dc-9606-adaa80fef0c7",
  "isActive": true,
  "isExpired": false,
  "isExternal": true,
  "issuer": null,
  "jwksUri": null,
  "modulus": null,
  "notBefore": 2024-01-01T08:00:00.000Z,
  "operations": [
    "sign",
    "verify",
  ],
  "ownerId": null,
  "purpose": null,
  "type": "OKP",
  "updatedAt": 2024-01-01T08:00:00.000Z,
  "use": "sig",
}
`;

exports[`KryptosKit (OKP) from auto jwk 2`] = `
{
  "alg": "EdDSA",
  "crv": "Ed25519",
  "iat": 1704096000,
  "key_ops": [
    "sign",
    "verify",
  ],
  "kid": "6fc93fb6-d40e-52dc-9606-adaa80fef0c7",
  "kty": "OKP",
  "nbf": 1704096000,
  "uat": 1704096000,
  "use": "sig",
  "x": "eTgk2sA8y7_9q8MjEPskpM4mAUZqtXmwEiHwy249kj4",
}
`;

exports[`KryptosKit (OKP) from auto jwk 3`] = `
{
  "alg": "EdDSA",
  "crv": "Ed25519",
  "d": "CDqMp2HAyaLiV8Xdt8ouEnhYbZtRXlFAsEkccUCcpQk",
  "kid": "6fc93fb6-d40e-52dc-9606-adaa80fef0c7",
  "kty": "OKP",
  "use": "sig",
  "x": "eTgk2sA8y7_9q8MjEPskpM4mAUZqtXmwEiHwy249kj4",
}
`;

exports[`KryptosKit (OKP) from auto pem 1`] = `
{
  "algorithm": "EdDSA",
  "createdAt": 2024-01-01T08:00:00.000Z,
  "curve": "Ed25519",
  "encryption": null,
  "expiresAt": null,
  "expiresIn": -1,
  "hasPrivateKey": true,
  "hasPublicKey": true,
  "hidden": false,
  "id": "3ee7c51b-e879-554e-ae49-e6330610f09e",
  "isActive": true,
  "isExpired": false,
  "isExternal": false,
  "issuer": null,
  "jwksUri": null,
  "modulus": null,
  "notBefore": 2024-01-01T08:00:00.000Z,
  "operations": [
    "sign",
    "verify",
  ],
  "ownerId": null,
  "purpose": null,
  "type": "OKP",
  "updatedAt": 2024-01-01T08:00:00.000Z,
  "use": "sig",
}
`;

exports[`KryptosKit (OKP) from auto pem 2`] = `
{
  "alg": "EdDSA",
  "crv": "Ed25519",
  "iat": 1704096000,
  "key_ops": [
    "sign",
    "verify",
  ],
  "kid": "3ee7c51b-e879-554e-ae49-e6330610f09e",
  "kty": "OKP",
  "nbf": 1704096000,
  "uat": 1704096000,
  "use": "sig",
  "x": "FqCTfRe3gmMUYSmAmAl8id-ZYSsmFpKHXmh-ZOaPOfM",
}
`;

exports[`KryptosKit (OKP) from auto pem 3`] = `
{
  "algorithm": "EdDSA",
  "curve": "Ed25519",
  "id": "3ee7c51b-e879-554e-ae49-e6330610f09e",
  "privateKey": "-----BEGIN PRIVATE KEY-----
MC4CAQAwBQYDK2VwBCIEIA5gn8ikyFzW8+jNHtl4CvQvLSC2xXJVKTqsdY49bwma
-----END PRIVATE KEY-----
",
  "publicKey": "-----BEGIN PUBLIC KEY-----
MCowBQYDK2VwAyEAFqCTfRe3gmMUYSmAmAl8id+ZYSsmFpKHXmh+ZOaPOfM=
-----END PUBLIC KEY-----
",
  "type": "OKP",
  "use": "sig",
}
`;

exports[`KryptosKit (OKP) from b64 1`] = `
{
  "algorithm": "EdDSA",
  "createdAt": 2024-01-01T08:00:00.000Z,
  "curve": "Ed25519",
  "encryption": null,
  "expiresAt": null,
  "expiresIn": -1,
  "hasPrivateKey": true,
  "hasPublicKey": true,
  "hidden": false,
  "id": "ff2a448e-8eb0-59a8-8fc6-0ce23ad07e8f",
  "isActive": true,
  "isExpired": false,
  "isExternal": false,
  "issuer": null,
  "jwksUri": null,
  "modulus": null,
  "notBefore": 2024-01-01T08:00:00.000Z,
  "operations": [
    "sign",
    "verify",
  ],
  "ownerId": null,
  "purpose": null,
  "type": "OKP",
  "updatedAt": 2024-01-01T08:00:00.000Z,
  "use": "sig",
}
`;

exports[`KryptosKit (OKP) from b64 2`] = `
{
  "alg": "EdDSA",
  "crv": "Ed25519",
  "iat": 1704096000,
  "key_ops": [
    "sign",
    "verify",
  ],
  "kid": "ff2a448e-8eb0-59a8-8fc6-0ce23ad07e8f",
  "kty": "OKP",
  "nbf": 1704096000,
  "uat": 1704096000,
  "use": "sig",
  "x": "GRCwCA6lChosFGMQwxGiHCdzblfvCz0FNiRtTnm1qqc",
}
`;

exports[`KryptosKit (OKP) from b64 3`] = `
{
  "algorithm": "EdDSA",
  "curve": "Ed25519",
  "id": "ff2a448e-8eb0-59a8-8fc6-0ce23ad07e8f",
  "privateKey": "MC4CAQAwBQYDK2VwBCIEIBwKJlvoh1ngd9LRd7dtvGOSqW4uZamdvIu0ABD2AkxL",
  "publicKey": "MCowBQYDK2VwAyEAGRCwCA6lChosFGMQwxGiHCdzblfvCz0FNiRtTnm1qqc",
  "type": "OKP",
  "use": "sig",
}
`;

exports[`KryptosKit (OKP) from jwk 1`] = `
{
  "algorithm": "EdDSA",
  "createdAt": 2024-01-01T08:00:00.000Z,
  "curve": "Ed25519",
  "encryption": null,
  "expiresAt": null,
  "expiresIn": -1,
  "hasPrivateKey": true,
  "hasPublicKey": true,
  "hidden": false,
  "id": "6fc93fb6-d40e-52dc-9606-adaa80fef0c7",
  "isActive": true,
  "isExpired": false,
  "isExternal": true,
  "issuer": null,
  "jwksUri": null,
  "modulus": null,
  "notBefore": 2024-01-01T08:00:00.000Z,
  "operations": [
    "sign",
    "verify",
  ],
  "ownerId": null,
  "purpose": null,
  "type": "OKP",
  "updatedAt": 2024-01-01T08:00:00.000Z,
  "use": "sig",
}
`;

exports[`KryptosKit (OKP) from jwk 2`] = `
{
  "alg": "EdDSA",
  "crv": "Ed25519",
  "iat": 1704096000,
  "key_ops": [
    "sign",
    "verify",
  ],
  "kid": "6fc93fb6-d40e-52dc-9606-adaa80fef0c7",
  "kty": "OKP",
  "nbf": 1704096000,
  "uat": 1704096000,
  "use": "sig",
  "x": "eTgk2sA8y7_9q8MjEPskpM4mAUZqtXmwEiHwy249kj4",
}
`;

exports[`KryptosKit (OKP) from jwk 3`] = `
{
  "alg": "EdDSA",
  "crv": "Ed25519",
  "d": "CDqMp2HAyaLiV8Xdt8ouEnhYbZtRXlFAsEkccUCcpQk",
  "kid": "6fc93fb6-d40e-52dc-9606-adaa80fef0c7",
  "kty": "OKP",
  "use": "sig",
  "x": "eTgk2sA8y7_9q8MjEPskpM4mAUZqtXmwEiHwy249kj4",
}
`;

exports[`KryptosKit (OKP) from pem 1`] = `
{
  "algorithm": "EdDSA",
  "createdAt": 2024-01-01T08:00:00.000Z,
  "curve": "Ed25519",
  "encryption": null,
  "expiresAt": null,
  "expiresIn": -1,
  "hasPrivateKey": true,
  "hasPublicKey": true,
  "hidden": false,
  "id": "3ee7c51b-e879-554e-ae49-e6330610f09e",
  "isActive": true,
  "isExpired": false,
  "isExternal": false,
  "issuer": null,
  "jwksUri": null,
  "modulus": null,
  "notBefore": 2024-01-01T08:00:00.000Z,
  "operations": [
    "sign",
    "verify",
  ],
  "ownerId": null,
  "purpose": null,
  "type": "OKP",
  "updatedAt": 2024-01-01T08:00:00.000Z,
  "use": "sig",
}
`;

exports[`KryptosKit (OKP) from pem 2`] = `
{
  "alg": "EdDSA",
  "crv": "Ed25519",
  "iat": 1704096000,
  "key_ops": [
    "sign",
    "verify",
  ],
  "kid": "3ee7c51b-e879-554e-ae49-e6330610f09e",
  "kty": "OKP",
  "nbf": 1704096000,
  "uat": 1704096000,
  "use": "sig",
  "x": "FqCTfRe3gmMUYSmAmAl8id-ZYSsmFpKHXmh-ZOaPOfM",
}
`;

exports[`KryptosKit (OKP) from pem 3`] = `
{
  "algorithm": "EdDSA",
  "curve": "Ed25519",
  "id": "3ee7c51b-e879-554e-ae49-e6330610f09e",
  "privateKey": "-----BEGIN PRIVATE KEY-----
MC4CAQAwBQYDK2VwBCIEIA5gn8ikyFzW8+jNHtl4CvQvLSC2xXJVKTqsdY49bwma
-----END PRIVATE KEY-----
",
  "publicKey": "-----BEGIN PUBLIC KEY-----
MCowBQYDK2VwAyEAFqCTfRe3gmMUYSmAmAl8id+ZYSsmFpKHXmh+ZOaPOfM=
-----END PUBLIC KEY-----
",
  "type": "OKP",
  "use": "sig",
}
`;

exports[`KryptosKit (OKP) generate auto 1`] = `
{
  "algorithm": "ECDH-ES+A128GCMKW",
  "createdAt": 2024-01-01T08:00:00.000Z,
  "curve": "P-256",
  "encryption": "A256GCM",
  "expiresAt": null,
  "expiresIn": -1,
  "hasPrivateKey": true,
  "hasPublicKey": true,
  "hidden": false,
  "id": "6e6f84b0-e125-5e3f-90ae-c65269668d98",
  "isActive": true,
  "isExpired": false,
  "isExternal": false,
  "issuer": null,
  "jwksUri": null,
  "modulus": null,
  "notBefore": 2024-01-01T08:00:00.000Z,
  "operations": [
    "deriveKey",
  ],
  "ownerId": null,
  "purpose": null,
  "type": "EC",
  "updatedAt": 2024-01-01T08:00:00.000Z,
  "use": "enc",
}
`;

exports[`KryptosKit (OKP) generate enc 1`] = `
{
  "algorithm": "ECDH-ES",
  "createdAt": 2024-01-01T08:00:00.000Z,
  "curve": "X25519",
  "encryption": "A256GCM",
  "expiresAt": null,
  "expiresIn": -1,
  "hasPrivateKey": true,
  "hasPublicKey": true,
  "hidden": false,
  "id": "6e6f84b0-e125-5e3f-90ae-c65269668d98",
  "isActive": true,
  "isExpired": false,
  "isExternal": false,
  "issuer": null,
  "jwksUri": null,
  "modulus": null,
  "notBefore": 2024-01-01T08:00:00.000Z,
  "operations": [
    "deriveKey",
  ],
  "ownerId": null,
  "purpose": null,
  "type": "OKP",
  "updatedAt": 2024-01-01T08:00:00.000Z,
  "use": "enc",
}
`;

exports[`KryptosKit (OKP) generate sig 1`] = `
{
  "algorithm": "EdDSA",
  "createdAt": 2024-01-01T08:00:00.000Z,
  "curve": "Ed448",
  "encryption": null,
  "expiresAt": null,
  "expiresIn": -1,
  "hasPrivateKey": true,
  "hasPublicKey": true,
  "hidden": false,
  "id": "6e6f84b0-e125-5e3f-90ae-c65269668d98",
  "isActive": true,
  "isExpired": false,
  "isExternal": false,
  "issuer": null,
  "jwksUri": null,
  "modulus": null,
  "notBefore": 2024-01-01T08:00:00.000Z,
  "operations": [
    "sign",
    "verify",
  ],
  "ownerId": null,
  "purpose": null,
  "type": "OKP",
  "updatedAt": 2024-01-01T08:00:00.000Z,
  "use": "sig",
}
`;
